File: AEDS-II/verde/TP02/03/Show.java
import java.io.*;
import java.util.*;

public class Show{

    private String SHOW_ID;
    private String TYPE;
    private String TITLE;
    private String DIRECTOR;
    private String[] CAST;
    private String COUNTRY;
    private String DATE_ADDED;
    private int RELEASE_YEAR;
    private String RATING;
    private String DURATION;
    private String[] LISTED_IN;

    public Show(){

        this.SHOW_ID = this.TYPE = this.TITLE = this.DIRECTOR = this.COUNTRY = this.DATE_ADDED = this.RATING = this.DURATION = "NaN";
        this.RELEASE_YEAR = 0;
        this.CAST = new String[]{"NaN"};
        this.LISTED_IN = new String[]{"NaN"};
    }

    public void read(String line){

        List<String> fields = new ArrayList<>();
        
        boolean insideQuotes = false;
        StringBuilder currentField = new StringBuilder();

        for(int i = 0; i < line.length(); i++){
            
            char c = line.charAt(i);
            
            if (c == '"'){
                insideQuotes = !insideQuotes;
            }
            else if (c == ',' && !insideQuotes){
            
                fields.add(currentField.toString());
                currentField.setLength(0);
            } 
            else{
                currentField.append(c);
            }
        }

        fields.add(currentField.toString());

        this.SHOW_ID = fields.get(0);
        this.TYPE = fields.get(1);
        this.TITLE = fields.get(2);
        this.DIRECTOR = fields.get(3).isEmpty() ? "NaN" : fields.get(3);
        this.CAST = fields.get(4).isEmpty() ? new String[]{"NaN"} : fields.get(4).split(";\\s*");
        this.COUNTRY = fields.get(5).isEmpty() ? "NaN" : fields.get(5);
        this.DATE_ADDED = fields.get(6).isEmpty() ? "NaN" : fields.get(6);
        this.RELEASE_YEAR = fields.get(7).isEmpty() ? 0 : Integer.parseInt(fields.get(7));
        this.RATING = fields.get(8).isEmpty() ? "NaN" : fields.get(8);
        this.DURATION = fields.get(9).isEmpty() ? "NaN" : fields.get(9);
        this.LISTED_IN = fields.get(10).isEmpty() ? new String[]{"NaN"} : fields.get(10).split(";\\s*");
    }

    public String getTITLE(){
        return TITLE;
    }

    public String toString(){

        return "=> " + SHOW_ID + " ## " + TITLE + " ## " + TYPE + " ## " + DIRECTOR + " ## " + Arrays.toString(CAST)
                + " ## " + COUNTRY + " ## " + DATE_ADDED + " ## " + RELEASE_YEAR + " ## " + RATING + " ## " + DURATION
                + " ## " + Arrays.toString(LISTED_IN) + " ##";
    }

    public static void main(String[] args) throws Exception{
    
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        
        String path = "/tmp/disneyplus.csv";
        BufferedReader csvReader = new BufferedReader(new FileReader(path));

        String line;
        List<Show> list = new ArrayList<>();

        csvReader.readLine(); 

        while((line = csvReader.readLine()) != null){
        
            Show show = new Show();
            show.read(line);
            list.add(show);
        }
        
        csvReader.close();

        Show[] vetor = new Show[3000];
        int n = 0;
        String input;

        while(!(input = in.readLine()).equals("FIM")){
            
            for(Show s : list){
                if(s.SHOW_ID.equals(input)){
                    
                    vetor[n++] = s;
                    break;
                }
            }
        }

        List<String> consultas = new ArrayList<>();

        while((input = in.readLine()) != null && !input.equals("FIM")){    
            consultas.add(input);
        }

        int comparacoes = 0;
        long inicio = System.nanoTime();

        for(String idConsulta : consultas){
            
            boolean encontrado = false;
            
            for(int i = 0; i < n; i++){   
                comparacoes++;
                
                if(vetor[i].SHOW_ID.equals(idConsulta)){
                    encontrado = true;
                    break;
                }
            }
            System.out.println(encontrado ? "SIM" : "NAO");
        }

        long fim = System.nanoTime();
        double tempo = (fim - inicio) / 1e6;

        PrintWriter log = new PrintWriter("matricula_sequencial.txt");

        log.printf("1234567\t%.3f\t%d\n", tempo, comparacoes);
        log.close();
    }
}

File: AEDS-II/verde/TP02/03/Show.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

#define MAX_LINE_SIZE 1024
#define MAX_FIELDS 12

typedef struct{
    char* SHOW_ID;
    char* TYPE;
    char* TITLE;
    char** DIRECTOR;
    int director_count;
    char** CAST;
    int cast_count;
    char* COUNTRY;
    char* DATE_ADDED;
    int RELEASE_YEAR;
    char* RATING;
    char* DURATION;
    char** LISTED_IN;
    int listed_in_count;
} Show;

Show* criarShow(){
    Show* show = (Show*)malloc(sizeof(Show));
    show->SHOW_ID = strdup("NaN");
    show->TYPE = strdup("NaN");
    show->TITLE = strdup("NaN");
    show->DIRECTOR = NULL;
    show->director_count = 0;
    show->CAST = NULL;
    show->cast_count = 0;
    show->COUNTRY = strdup("NaN");
    show->DATE_ADDED = strdup("NaN");
    show->RELEASE_YEAR = -1;
    show->RATING = strdup("NaN");
    show->DURATION = strdup("NaN");
    show->LISTED_IN = NULL;
    show->listed_in_count = 0;
    return show;
}

void liberarShow(Show* show){
    free(show->SHOW_ID);
    free(show->TYPE);
    free(show->TITLE);
    
    for (int i = 0; i < show->director_count; i++){
        free(show->DIRECTOR[i]);
    }
    free(show->DIRECTOR);
    
    for (int i = 0; i < show->cast_count; i++){
        free(show->CAST[i]);
    }
    free(show->CAST);
    
    free(show->COUNTRY);
    free(show->DATE_ADDED);
    free(show->RATING);
    free(show->DURATION);
    
    for (int i = 0; i < show->listed_in_count; i++){
        free(show->LISTED_IN[i]);
    }
    free(show->LISTED_IN);
    
    free(show);
}

char** splitString(const char* str, const char* delim, int* count){
    if (str == NULL || strlen(str) == 0){
        *count = 0;
        return NULL;
    }
    
    char* copy = strdup(str);
    char* token = strtok(copy, delim);
    char** result = NULL;
    int size = 0;
    
    while (token != NULL){
        result = (char**)realloc(result, (size + 1) * sizeof(char*));
        result[size] = strdup(token);
        size++;
        token = strtok(NULL, delim);
    }
    
    free(copy);
    *count = size;
    return result;
}

Show* parseLinha(char* linha){
    char* campos[MAX_FIELDS];
    int campoIndex = 0;
    bool dentroAspas = false;
    char* inicioCampo = linha;
    
    for (char* p = linha; *p; p++){
        if (*p == '"'){
            dentroAspas = !dentroAspas;
        } else if (*p == ',' && !dentroAspas){
            *p = '\0';
            campos[campoIndex++] = inicioCampo;
            inicioCampo = p + 1;
        }
    }
    campos[campoIndex] = inicioCampo;
    
    Show* show = criarShow();
    
    if (campoIndex >= 0) show->SHOW_ID = strdup(campos[0]);
    if (campoIndex >= 1) show->TYPE = strdup(campos[1]);
    if (campoIndex >= 2) show->TITLE = strdup(campos[2]);
    
    if (campoIndex >= 3){
        show->DIRECTOR = splitString(campos[3], ", ", &show->director_count);
    }
    
    if (campoIndex >= 4){
        show->CAST = splitString(campos[4], ", ", &show->cast_count);
    }
    
    if (campoIndex >= 5) show->COUNTRY = strdup(campos[5]);
    if (campoIndex >= 6) show->DATE_ADDED = strdup(campos[6]);
    
    if (campoIndex >= 7){
        show->RELEASE_YEAR = atoi(campos[7]);
    }
    
    if (campoIndex >= 8) show->RATING = strdup(campos[8]);
    if (campoIndex >= 9) show->DURATION = strdup(campos[9]);
    
    if (campoIndex >= 10){
        show->LISTED_IN = splitString(campos[10], ", ", &show->listed_in_count);
    }
    
    return show;
}

int main(){
    clock_t inicio = clock();
    int comparacoes = 0;
    
    FILE* arquivo = fopen("/tmp/disneyplus.csv", "r");
    if (arquivo == NULL){
        perror("Erro ao abrir arquivo");
        return 1;
    }
    
    char linha[MAX_LINE_SIZE];
    fgets(linha, sizeof(linha), arquivo);
    
    Show** programas = NULL;
    int numProgramas = 0;
    
    while (fgets(linha, sizeof(linha), arquivo)){
        
        linha[strcspn(linha, "\n")] = '\0';
        Show* programa = parseLinha(linha);
        programas = (Show**)realloc(programas, (numProgramas + 1) * sizeof(Show*));
        programas[numProgramas++] = programa;
    }
    fclose(arquivo);
    
    char entrada[MAX_LINE_SIZE];
    scanf("%[^\n]%*c", entrada);
    
    while (strcmp(entrada, "FIM") != 0){
        bool encontrado = false;
        
        for (int i = 0; i < numProgramas; i++){
            comparacoes++;
            if (programas[i]->TITLE && strcmp(programas[i]->TITLE, entrada) == 0){
                encontrado = true;
                break;
            }
        }
        
        printf("%s\n", encontrado ? "SIM" : "NAO");
        scanf("%[^\n]%*c", entrada);
    }
    
    for (int i = 0; i < numProgramas; i++){
        liberarShow(programas[i]);
    }
    free(programas);
    
    clock_t fim = clock();
    double duracao = ((double)(fim - inicio)) / CLOCKS_PER_SEC * 1000;
    
    FILE* log = fopen("123456_sequencial.txt", "w");
    if (log != NULL){
        fprintf(log, "123456\t%.0f\t%d\n", duracao, comparacoes);
        fclose(log);
    }
    
    return 0;
}